flink/src/main/java/org/apache/iceberg/flink/sink/FlinkFileAppenderFactory.java

```diff
@@ -0,0 +1,201 @@
+        file.encryptingOutputFile().location(), spec, partition, file.keyMetadata());
+  }
+
+  @Override
+  public EqualityDeleteWriter<RowData> newEqDeleteWriter(EncryptedOutputFile outputFile, FileFormat format,
+                                                         StructLike partition) {
+    MetricsConfig metricsConfig = MetricsConfig.fromProperties(props);
+    try {
+      switch (format) {
+        case AVRO:
+          return Avro.writeDeletes(outputFile.encryptingOutputFile())
+              .createWriterFunc(ignore -> new FlinkAvroWriter(flinkSchema))
+              .withPartition(partition)
+              .overwrite()
+              .setAll(props)
+              .rowSchema(schema)
+              .withSpec(spec)
+              .withKeyMetadata(outputFile.keyMetadata())
+              .equalityFieldIds(equalityFieldIds)
+              .buildEqualityWriter();
+
+        case PARQUET:
+          return Parquet.writeDeletes(outputFile.encryptingOutputFile())
+              .createWriterFunc(msgType -> FlinkParquetWriters.buildWriter(flinkSchema, msgType))
+              .withPartition(partition)
+              .overwrite()
+              .setAll(props)
+              .metricsConfig(metricsConfig)
+              .rowSchema(schema)
+              .withSpec(spec)
+              .withKeyMetadata(outputFile.keyMetadata())
+              .equalityFieldIds(equalityFieldIds)
+              .buildEqualityWriter();
+
+        default:
+          throw new UnsupportedOperationException("Cannot write unknown file format: " + format);
+      }
+    } catch (IOException e) {
+      throw new UncheckedIOException(e);
+    }
+  }
+
+  @Override
+  public PositionDeleteWriter<RowData> newPosDeleteWriter(EncryptedOutputFile outputFile, FileFormat format,
+                                                          StructLike partition) {
+    MetricsConfig metricsConfig = MetricsConfig.fromProperties(props);
+    Schema rowSchema = posDeleteWithRow ? schema : null;
+    try {
+      switch (format) {
+        case AVRO:
+          Function<org.apache.avro.Schema, DatumWriter<?>> writeFunc = null;
+          if (rowSchema != null) {
+            writeFunc = ignore -> new FlinkAvroWriter(FlinkSchemaUtil.convert(rowSchema));
+          }
+
+          return Avro.writeDeletes(outputFile.encryptingOutputFile())
+              .createWriterFunc(writeFunc)
+              .withPartition(partition)
+              .overwrite()
+              .setAll(props)
+              .rowSchema(rowSchema)
+              .withSpec(spec)
+              .withKeyMetadata(outputFile.keyMetadata())
+              .buildPositionWriter();
+
+        case PARQUET:
+          RowType flinkParquetRowType = FlinkSchemaUtil.convert(DeleteUtil.posDeleteSchema(rowSchema));
+
+          return Parquet.writeDeletes(outputFile.encryptingOutputFile())
+              .createWriterFunc(msgType -> FlinkParquetWriters.buildWriter(flinkParquetRowType, msgType))
```

Comment:
I don't think this will work because the writer that is returned will be wrapped by a `PositionDeleteStructWriter`. That would duplicate the position delete struct because this is going to produce a writer for it as well. That's why the schema passed here is a bug, like I noted above.

Reply:
Yeah, this needs to be fixed. I'm looking through how Avro actually works right now and it is okay because we're calling `setSchema` a second time from the position writer, which basically discards the original writer with the extra record and rebuilds it with just the row schema.

I'll open a PR to fix it.